### PR TITLE
feat(developer): loca minimization + cleanup kmldmlc error list 🙀

### PR DESCRIPTION
**Pull request by mcdurdin**
Yeah two for the price of one:

- minimize locales at compile time
- cleans up the list of errors to make it easier to maintain and use

TODO -- on follow-up PRs:

- Add kmldmlc test to developer test build -> see #7180
- Ensure that DTD is used when verifying XML files -> see #7181

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/kmldmlc/src/keyman/compiler/loca.ts [hunk 1]
```diff
@@ -35,18 +40,20 @@ export class LocaCompiler extends SectionCompiler {
...
-    let locales = this.getLocales(this.keyboard);
+    // This also minimizes locales according to Remove Likely Subtags algorithm:
+    // https://www.unicode.org/reports/tr35/#Likely_Subtags
+    let locales = this.getLocales(this.keyboard).map((locale: string) => new Intl.Locale(locale).minimize().toString());
+
     // TODO: remove `as any` cast: (Intl as any): ts lib version we have doesn't
     // yet include `getCanonicalLocales` but node 16 does include it so we can
     // safely use it. Also well supported in modern browsers.
-    result.locales = (Intl as any).getCanonicalLocales(locales);
+    result.locales = ((Intl as any).getCanonicalLocales(locales));
```
Comment (mcdurdin) [inline developer/src/kmldmlc/src/keyman/compiler/loca.ts:50]:
```suggestion
    result.locales = (Intl as any).getCanonicalLocales(locales);
```

File: developer/src/kmldmlc/src/keyman/compiler/loca.ts [hunk 2]
```diff
@@ -35,18 +40,20 @@ export class LocaCompiler extends SectionCompiler {
...
 
     if(result.locales.length < locales.length) {
       // TODO-LDML: hint on repeated locales
+      this.callbacks.reportMessage(CompilerErrors.OneOrMoreRepeatedLocales())
```
Comment (mcdurdin) [inline developer/src/kmldmlc/src/keyman/compiler/loca.ts:54]:
```suggestion
      this.callbacks.reportMessage(CompilerErrors.OneOrMoreRepeatedLocales());
```

File: developer/src/kmldmlc/src/keyman/compiler/errors.ts [hunk 3]
```diff
@@ -9,19 +9,44 @@ export enum CompilerErrorSeverity {
...
+  static RowOnHardwareLayerHasTooManyKeys = (o:{row: number}) =>  m(this.ERROR_RowOnHardwareLayerHasTooManyKeys, `Row #${o.row} on 'hardware' layer has too many keys`);
+  static ERROR_RowOnHardwareLayerHasTooManyKeys = SevError | 0x0004;
+
+  static KeyNotFoundInKeyBag = (o:{keyId: string, col: number, row: number, layer: string, form: string}) =>
+     m(this.ERROR_KeyNotFoundInKeyBag, `Key ${o.keyId} in position #${o.col} on row #${o.row} of layer ${o.layer}, form '${o.form}' not found in key bag`);
+  static ERROR_KeyNotFoundInKeyBag = SevError | 0x0005;
+
+  static OneOrMoreRepeatedLocales = () =>
+    m(this.HINT_OneOrMoreRepeatedLocales, `After minimization, one or more locales is repeated and has been removed`);
+  static HINT_OneOrMoreRepeatedLocales = SevHint | 0x0006;
```
Comment (darcywong00) [inline developer/src/kmldmlc/src/keyman/compiler/errors.ts:38]: Confirming you're intending to include Hints in `CompilerErrors`.
Is it possible later on to also have `SevError | 0x0006;`?

Comment (mcdurdin) [reply developer/src/kmldmlc/src/keyman/compiler/errors.ts:38]: 1. Yes, all classes of 'error' from info through to fatal. Perhaps we should call it `CompilerMessages`?

2. Each code should be unique independent of its severity code, so we wouldn't use `SevError | 0x0006` in future. (In theory we may want to change the severity of a code one day?)

File: developer/src/kmldmlc/src/keyman/compiler/keys.ts [hunk 4]
```diff
@@ -36,7 +37,7 @@ export class KeysCompiler extends SectionCompiler {
     for(let row of layer.row) {
       y++;
       if(y > USVirtualKeyMap.length) {
-        this.callbacks.reportMessage(0, `'hardware' layer has too many rows`);
+        this.callbacks.reportMessage(CompilerErrors.HardwareLayerHasTooManyRows());
```
Comment (srl295) [inline developer/src/kmldmlc/src/keyman/compiler/keys.ts:40]: nice

File: developer/src/kmldmlc/test/test-loca.ts [hunk 5]
```diff
@@ -18,13 +18,16 @@ describe('loca', function () {
...
+    assert.equal(callbacks.messages.length, 1);
+    assert.deepEqual(callbacks.messages[0], CompilerErrors.OneOrMoreRepeatedLocales());
+
+    // Original is 6 locales, now five minimized in the results
     assert.equal(loca.locales.length, 5);
     assert.equal(loca.locales[0], 'mt');
-    assert.equal(loca.locales[1], 'fr-FR');
-    assert.equal(loca.locales[2], 'km-Khmr-KH');
+    assert.equal(loca.locales[1], 'fr');  // Original fr-FR
+    assert.equal(loca.locales[2], 'km');  // Original km-Khmr-kh
```
Comment (srl295) [inline developer/src/kmldmlc/test/test-loca.ts:30]:
```suggestion
    assert.equal(loca.locales[2], 'km');  // Original km-Khmr-KH
```

**Review by srl295 — APPROVED** [review]

File: developer/src/kmldmlc/test/test-loca.ts (hunk 5 above)
Comment (mcdurdin) [reply developer/src/kmldmlc/test/test-loca.ts:30]: See https://github.com/keymanapp/keyman/blob/7d38e26b09dcf06e00f316cd54cfa9a2fe4b41e1/developer/src/kmldmlc/test/fixtures/sections/loca/multiple.xml#L7 :) it's lower case in the source file!

Comment (srl295) [reply developer/src/kmldmlc/test/test-loca.ts:30]: oh ok.  i was going by the above line.

**Pull request merged**